design/external-references.md

```diff
@@ -0,0 +1,394 @@
+- **REQ5**: If a resource is exported twice under the same export name, only a single set of outputs will be created ([#1496](https://github.com/awslabs/aws-cdk/issues/1496)).
+- **REQ6**: If the same resource is imported twice in the same stack, the "import" method will return the same object.
+- **REQ7**: For some resources (like VPC for example), it should be possible look up the resource in the current environment by querying through an environmental context provider.
+- **REQ8**: It should be possible to resolve import values either during synthesis time (i.e. via an environmental context provider) or during deploy time (i.e. via `Fn::ImportValue`). There are certain resources that will _require_ the use of synthesis-time resolution due to their complex representation.
+
+## Approach
+
+As usual, we'll take a layered approach.
+
+**Serializable constructs**: At the low-level, we'll define what it means for a construct to be "__serializable__" through string key/value context. The mechanism will be recursive such that it will be possible to write/read singular values and write/read serializable objects (which, themselves will write/read their own values and so forth).
+
+**Serialization context**: the notion of a "string key/value context" represents the lowest common denominator for serialization, and specifically is the only type supported by CloudFormation's cross-stack export/import mechanism (CloudFormation outputs can only be of "string" type). If a construct can serialize itself through a set of key/values, we can pass it via CloudFormation's import/export, environment variables, SSM parameters, etc.
+
+**CloudFormation imports/exports as a serialization context**: on top of that, we will implement a serialization/deserialization context based on AWS CloudFormation's import/exports mechanism. "writing" a value means defining an Output with an Export and "reading" a value means `Fn::ImportValue` with this export name. The "export name" will be used as a prefix that represent an object, and sub-objects will be serialized by adding another component to the export name.
+
+**Synthesis-time Imports**: we will also define an environmental context provider that will import values during synthesis and store them in `cdk.json`. This approach is more robust since it will allow CDK code to reason about concrete values instead of opaque tokens. It's especially needed for situations where a list of values needs to be passed and the arity is required.
+
+**Convenience methods**: now that constructs can be serializable through imports/exports, we can implement a set of convenience methods for each AWS resource to provide nice ergonomics for the specific use case of exporting and importing resources across apps.
+
+## Design
+
+### Serialization
+
+If a construct is said to be serializable, it must implement the `ISerializable` interface:
```

Comment:
Interface name is slippery slope, risks confusion & collisions. I'm not *extremely* concerned about it, but if we can come up with a non-conflicting name I think that's better off.